src/toil/leader.py

```diff
@@ -322,6 +352,126 @@ def __init__( self, jobStoreString, numberOfFailedJobs ):
+        :return: a jobWrapper added to scheduleServices whose services are running, or None if 
+        no such job is available.
+        """
+        try:
+            return self._jobWrappersWithServicesThatHaveStarted.get(timeout=maxWait)
+        except Empty:
+            return None
+    
+    def getServiceJobsToStart(self, maxWait):
+        """
+        :param string maxWait: Time in seconds to wait to get a jobWrapper before returning.
+        :return: a tuple of (serviceJobStoreID, memory, cores, disk, predecessorID) representing
+        a service job to start.
+        """
+        try:
+            return self._serviceJobWrappersToStart.get(timeout=maxWait)
+        except Empty:
+            return None     
+
+    def killServices(self, jobWrapper):
+        """
+        :param toil.jobWrapper.jobWrapper jobWrapper: Kill the services of this
+        job asynchronously.
+        """
+        for serviceTuple in jobWrapper.services:
+            self.jobStore.delete(serviceTuple[5])
+        
+    def shutdown(self):
+        """
+        Cleanly terminate worker threads starting and killing services. Will block
+        until all services are started and blocked.
+        """
+        self._terminate.set()
+        self._serviceStarter.join()
+    
+    @staticmethod
+    def _startServices(jobWrappersWithServicesToStart, 
+                       jobWrappersWithServicesThatHaveStarted, 
+                       serviceJobsToStart,
+                       jobStore, terminate):
+        """
+        Thread used to schedule services.
+        """
+        while True:
+            try:
+                # Get a jobWrapper with services to start, waiting a short period
+                jobWrapper = jobWrappersWithServicesToStart.get(1.0) 
+            except:
+                # Check if the thread should quit
+                if terminate.is_set():
+                    logger.debug('Received signal to quit starting services.')
+                    break
+                continue
+                
+            if jobWrapper is None: # Nothing was ready, loop again
+                continue
+            
+            # Start the service jobs in batches, waiting for each batch
+            # to become established before starting the next batch
+            for serviceJobList in jobWrapper.services: 
+                for serviceJobStoreID, memory, cores, disk, startJobStoreID, terminateJobStoreID in serviceJobList:                
+                    assert jobStore.exists(startJobStoreID)
+                    assert jobStore.exists(terminateJobStoreID)
+                    serviceJobsToStart.put((serviceJobStoreID, memory, cores, disk))
+                
+                # Wait until all the services of the batch are running
+                for serviceTuple in serviceJobList:
+                    while jobStore.exists(serviceTuple[4]):
+                        # Sleep to avoid thrashing
+                        time.sleep(1.0) 
```

Comment:
This should use exponential back-off, see #557. We will hit the request rate limit if we are not careful!!!